File: 2059-unique-length-3-palindromic-subsequences/2059-unique-length-3-palindromic-subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        set<char> chars(s.begin(), s.end());
        int ans = 0;
        for (auto ch : chars) {
            int firstPos = -1, lastPos = -1;
            for (int i = 0; i < s.size(); i++) {
                if (ch == s[i]) {
                    if (firstPos == -1) {
                        firstPos = i;
                    }
                    lastPos = i;
                }
            }

            set<char> mid;
            for (int k = firstPos + 1; k < lastPos; k++) {
                mid.insert(s[k]);
            }
            ans += mid.size();
        }

        return ans;
    }
};

File: 2059-unique-length-3-palindromic-subsequences/2059-unique-length-3-palindromic-subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        s = list(s)
        start = Counter()
        for i, ch in enumerate(s):
            if ch not in start:
                start[ch] = i

        end = Counter()
        for i in range(len(s) - 1, -1, -1):
            ch = s[i]
            if ch not in end:
                end[ch] = i

        ans = 0
        for i in range(97, 123):
            ch = chr(i)
            f, e = start[ch], end[ch]
            st = set()
            for j in range(f + 1, e):
                st.add(s[j])
            ans += len(st)
        return ans
